File: CloudBackup.Model/UserPermissions.cs

namespace CloudBackup.Model
{
    public class UserPermissions
    {
        /// <summary>
        /// If role represents global admin. Global admins have access to administrative menu.
        /// They also have access to other tenant's users.
        /// </summary>
        public bool IsGlobalAdmin { get; set; }

        /// <summary>
        /// If role represents user admin. User admins have access to administrative menu.
        /// </summary>
        public bool IsUserAdmin { get; set; }

        public Permissions ProfileRights { get; set; }

        public Permissions TenantRights { get; set; }

        public Permissions UserRights { get; set; }

        public Permissions InfrastructureRights { get; set; }

        public Permissions BackupRights { get; set; }

        public Permissions RestoreRights { get; set; }

        public Permissions SecurityRights { get; set; }

        public UserPermissions Combine(UserPermissions other)
        {
            var result = Clone();

            if (other != null)
            {
                result.IsGlobalAdmin |= other.IsGlobalAdmin;
                result.IsUserAdmin |= other.IsUserAdmin;
                result.ProfileRights |= other.ProfileRights;
                result.TenantRights |= other.TenantRights;
                result.UserRights |= other.UserRights;
                result.InfrastructureRights |= other.InfrastructureRights;
                result.BackupRights |= other.BackupRights;
                result.RestoreRights |= other.RestoreRights;
                result.SecurityRights |= other.SecurityRights;
            }

            return result;
        }

        public static UserPermissions Combine(UserPermissions permissions1, UserPermissions permissions2)
        {
            if (permissions1 == null)
                throw new ArgumentNullException(nameof(permissions1));

            return permissions1.Combine(permissions2);
        }

        public UserPermissions Clone()
        {
            return new UserPermissions
            {
                IsGlobalAdmin = IsGlobalAdmin,
                IsUserAdmin = IsUserAdmin,
                ProfileRights = ProfileRights,
                TenantRights = TenantRights,
                UserRights = UserRights,
                InfrastructureRights = InfrastructureRights,
                BackupRights = BackupRights,
                RestoreRights = RestoreRights,
                SecurityRights = SecurityRights
            };
        }
    }

    [Flags]
    public enum Permissions : byte
    {
        None = 0x0,
        Read = 0x1,
        Write = 0x2,
        ReadWrite = Read | Write
    }
}

File: CloudBackup.Management/ClientApp/src/app/classes/UserPermissions.ts
export enum CustomPermissions {
    None = 0,
    Read = 1 << 0,
    Write = 1 << 1,
    ReadWrite = Read | Write
}

export class UserPermissions {
    public IsGlobalAdmin = false;
    public IsUserAdmin = false;
    public ProfileRights = CustomPermissions.None;
    public TenantRights = CustomPermissions.None;
    public UserRights = CustomPermissions.None;
    public InfrastructureRights = CustomPermissions.None;
    public BackupRights = CustomPermissions.None;
    public RestoreRights = CustomPermissions.None;
    public SecurityRights = CustomPermissions.None;


    static Copy(obj: any): UserPermissions {
        let copy = new UserPermissions();
        copy.IsGlobalAdmin = obj.IsGlobalAdmin || obj.isGlobalAdmin;
        copy.IsUserAdmin = obj.IsUserAdmin || obj.isUserAdmin;
        copy.ProfileRights = obj.ProfileRights || obj.profileRights || CustomPermissions.None;
        copy.TenantRights = obj.TenantRights || obj.tenantRights || CustomPermissions.None;
        copy.UserRights = obj.UserRights || obj.userRights || CustomPermissions.None;
        copy.InfrastructureRights = obj.InfrastructureRights || obj.infrastructureRights || CustomPermissions.None;
        copy.BackupRights = obj.BackupRights || obj.backupRights || CustomPermissions.None;
        copy.RestoreRights = obj.RestoreRights || obj.restoreRights || CustomPermissions.None;
        copy.SecurityRights = obj.SecurityRights || obj.securityRights || CustomPermissions.None;
        return copy;
    }
}